src/rust/engine/fs/store/src/remote.rs

```diff
@@ -57,17 +62,66 @@ enum ByteStoreError {
   Other(String),
 }
 
+impl ByteStoreError {
+  fn retryable(&self) -> bool {
+    match self {
+      ByteStoreError::Grpc(status) => status_is_retryable(status),
+      ByteStoreError::Other(_) => false,
+    }
+  }
+}
+
+impl From<Status> for ByteStoreError {
+  fn from(status: Status) -> ByteStoreError {
+    ByteStoreError::Grpc(status)
+  }
+}
+
+impl From<String> for ByteStoreError {
+  fn from(string: String) -> ByteStoreError {
+    ByteStoreError::Other(string)
+  }
+}
+impl From<std::io::Error> for ByteStoreError {
+  fn from(err: std::io::Error) -> ByteStoreError {
+    ByteStoreError::Other(err.to_string())
+  }
+}
+
 impl fmt::Display for ByteStoreError {
   fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
     match self {
-      ByteStoreError::Grpc(status) => fmt::Display::fmt(status, f),
+      ByteStoreError::Grpc(status) => fmt::Display::fmt(&status_ref_to_str(status), f),
       ByteStoreError::Other(msg) => fmt::Display::fmt(msg, f),
     }
   }
 }
 
 impl std::error::Error for ByteStoreError {}
 
+/// Places that write the result of a remote `load`
+#[async_trait]
+trait LoadDestination: AsyncWrite + Send + Sync + Unpin + 'static {
```

Comment:
Nit: it seems like rather than a `reset` method, this could potentially be generic in terms of `new() -> Self` instead, and create a new `tempfile` or new `Vec`? I think that that would avoid needing to put this in a `Mutex`, because you could create it on the stack before the stream attempt.

But in general: I like this interface.

Reply:
Yeah, mutating to reinitialise via `reset` is more than a bit awkward, especially with how it forces the `Arc<Mutex<>>` overhead 😦.

In writing this comment, I discovered that the `load_monomorphic` call can at least just be `&mut dyn LoadDestination`, which makes the implementation of `load` much nicer, and avoids 'infecting' the interface, so I've made that adjustment. (However, due to the `retry_call` in the body, the current REAPI implementation needs to wrap that `&mut` in an `Arc<Mutex<>>` still, so that it can be cloned for each call. I suspect this can be improved, but I assume it'd require adjusting `retry_call` somehow, and I haven't worked that out yet...)

Having `new() -> Self` would be a possibility, but I don't think it works with my desired next steps for this code: separating the actually-downloading parts of `load_monomorphic` into a separate trait to support swapping out the remote store for https://github.com/pantsbuild/pants/pull/17840. Something like:

```rust
#[async_trait]
pub trait ByteStoreProvider: Sync + Send + 'static {
  async fn store_bytes(&self, digest: Digest, bytes: ByteSource) -> Result<(), String>;
  async fn load(&self, digest: Digest, destination: Arc<Mutex<dyn LoadDestination>>) -> Result<bool, String>;
}
```

That trait is best used behind a trait object (`Arc<dyn ByteStoreProvider>`), and thus `load` needs to be object-safe/cannot be generic. That is, it cannot be `async fn load<D: LoadDestination>(...)` to allow calling `D::new()` internally.

(Even without that goal, it's nice for compilation time and binary size for the complicated/significant `load_monomorphic` code to be non-generic so that it's only compiled once, rather than monomorphised for each different `LoadDestination` parameter it is used with.)